// chat-form/dist/umd/Message.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "prop-types", "react", "./App.module.scss", "./assets/description.svg"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("prop-types"), require("react"), require("./App.module.scss"), require("./assets/description.svg"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.propTypes, global.react, global.AppModule, global.description);
    global.undefined = mod.exports;
  }
})(this, function (exports, _propTypes, _react, _AppModule, _description) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _react2 = _interopRequireDefault(_react);

  var _AppModule2 = _interopRequireDefault(_AppModule);

  var _description2 = _interopRequireDefault(_description);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const MessageView = props => {
    const message = props;
    const [formattedTime, setFormatedTime] = (0, _react.useState)('');
    (0, _react.useEffect)(() => {
      import('dayjs').then(({
        default: moment
      }) => {
        setFormatedTime(moment(message.time).format('HH:mm'));
      });
    }, [message.time]);
    return /*#__PURE__*/_react2.default.createElement("div", {
      "data-testid": "message",
      style: {
        minHeight: 32
      }
    }, /*#__PURE__*/_react2.default.createElement("div", {
      onClick: message.canClick,
      className: [message.sender === 'bot' ? _AppModule2.default.botText : _AppModule2.default.inputText, _AppModule2.default.text].join(' ')
    }, message.message && /*#__PURE__*/_react2.default.createElement("span", null, message.message), message.fileSrc && /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, ['image/png', 'image/jpg'].includes(message?.file?.type || '') && /*#__PURE__*/_react2.default.createElement("img", {
      alt: "",
      src: message.fileSrc
    }), ['video/mp4'].includes(message?.file?.type || '') && /*#__PURE__*/_react2.default.createElement("video", {
      src: message.fileSrc
    }), !['video/mp4', 'image/png', 'image/jpg'].includes(message?.file?.type || '') && /*#__PURE__*/_react2.default.createElement("div", {
      className: _AppModule2.default.attachment
    }, /*#__PURE__*/_react2.default.createElement("div", {
      className: _AppModule2.default.image
    }, /*#__PURE__*/_react2.default.createElement("img", {
      alt: "",
      src: _description2.default
    })), /*#__PURE__*/_react2.default.createElement("div", {
      className: _AppModule2.default.fileName
    }, message?.file?.name))), /*#__PURE__*/_react2.default.createElement("span", {
      className: _AppModule2.default.time
    }, formattedTime)), /*#__PURE__*/_react2.default.createElement("br", null));
  };

  MessageView.propTypes = {
    message: _propTypes2.default.string.isRequired,
    sender: _propTypes2.default.oneOf(['bot', 'user']).isRequired,
    time: _propTypes2.default.string,
    canClick: _propTypes2.default.func,
    fileSrc: _propTypes2.default.string
  };
  exports.default = MessageView;
});
