// chat-form/dist/umd/Options.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "prop-types", "react", "./App.module.scss"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("prop-types"), require("react"), require("./App.module.scss"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.propTypes, global.react, global.AppModule);
    global.undefined = mod.exports;
  }
})(this, function (exports, _propTypes, _react, _AppModule) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _react2 = _interopRequireDefault(_react);

  var _AppModule2 = _interopRequireDefault(_AppModule);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const Options = props => {
    const {
      options,
      onSelect,
      selected,
      onFinish
    } = props;
    const [search, setSearch] = (0, _react.useState)('');

    const onSearch = e => {
      e.persist();
      setSearch(e.target.value);
    };

    return /*#__PURE__*/_react2.default.createElement("div", {
      className: _AppModule2.default.options
    }, /*#__PURE__*/_react2.default.createElement("input", {
      style: {
        backgroundColor: '#fff',
        width: '100%',
        marginBottom: 16,
        height: 24
      },
      className: _AppModule2.default.input,
      onChange: onSearch,
      placeholder: "Search"
    }), /*#__PURE__*/_react2.default.createElement("div", {
      style: {
        maxHeight: 160,
        overflowY: 'scroll'
      }
    }, options.filter(option => option.toLowerCase().includes(search.toLowerCase())).map(option => /*#__PURE__*/_react2.default.createElement("div", {
      key: option
    }, /*#__PURE__*/_react2.default.createElement("button", {
      onClick: () => onSelect(option),
      className: [_AppModule2.default.option, selected.includes(option) ? _AppModule2.default.selected : ''].join(' ')
    }, option)))), /*#__PURE__*/_react2.default.createElement("div", {
      style: {
        display: 'flex',
        justifyContent: 'flex-end'
      }
    }, /*#__PURE__*/_react2.default.createElement("button", {
      onClick: onFinish,
      className: _AppModule2.default.finish
    }, "Finish")));
  };

  Options.propTypes = {
    options: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired,
    onSelect: _propTypes2.default.func.isRequired,
    onFinish: _propTypes2.default.func.isRequired,
    selected: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired
  };
  exports.default = Options;
});
