// chat-form/dist/js/Message.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _AppModule = _interopRequireDefault(require("./App.module.scss"));

var _description = _interopRequireDefault(require("./assets/description.svg"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var MessageView = function MessageView(props) {
  var _message$file, _message$file2, _message$file3, _message$file4;

  var message = props;

  var _useState = (0, _react.useState)(''),
      _useState2 = _slicedToArray(_useState, 2),
      formattedTime = _useState2[0],
      setFormatedTime = _useState2[1];

  (0, _react.useEffect)(function () {
    Promise.resolve().then(function () {
      return _interopRequireWildcard(require('dayjs'));
    }).then(function (_ref) {
      var moment = _ref.default;
      setFormatedTime(moment(message.time).format('HH:mm'));
    });
  }, [message.time]);
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": "message",
    style: {
      minHeight: 32
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    onClick: message.canClick,
    className: [message.sender === 'bot' ? _AppModule.default.botText : _AppModule.default.inputText, _AppModule.default.text].join(' ')
  }, message.message && /*#__PURE__*/_react.default.createElement("span", null, message.message), message.fileSrc && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, ['image/png', 'image/jpg'].includes((message === null || message === void 0 ? void 0 : (_message$file = message.file) === null || _message$file === void 0 ? void 0 : _message$file.type) || '') && /*#__PURE__*/_react.default.createElement("img", {
    alt: "",
    src: message.fileSrc
  }), ['video/mp4'].includes((message === null || message === void 0 ? void 0 : (_message$file2 = message.file) === null || _message$file2 === void 0 ? void 0 : _message$file2.type) || '') && /*#__PURE__*/_react.default.createElement("video", {
    src: message.fileSrc
  }), !['video/mp4', 'image/png', 'image/jpg'].includes((message === null || message === void 0 ? void 0 : (_message$file3 = message.file) === null || _message$file3 === void 0 ? void 0 : _message$file3.type) || '') && /*#__PURE__*/_react.default.createElement("div", {
    className: _AppModule.default.attachment
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _AppModule.default.image
  }, /*#__PURE__*/_react.default.createElement("img", {
    alt: "",
    src: _description.default
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _AppModule.default.fileName
  }, message === null || message === void 0 ? void 0 : (_message$file4 = message.file) === null || _message$file4 === void 0 ? void 0 : _message$file4.name))), /*#__PURE__*/_react.default.createElement("span", {
    className: _AppModule.default.time
  }, formattedTime)), /*#__PURE__*/_react.default.createElement("br", null));
};

MessageView.propTypes = {
  message: _propTypes.default.string.isRequired,
  sender: _propTypes.default.oneOf(['bot', 'user']).isRequired,
  time: _propTypes.default.string,
  canClick: _propTypes.default.func,
  fileSrc: _propTypes.default.string
};
var _default = MessageView;
exports.default = _default;
