// chat-form/dist/umd/__tests__/ChatBot.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["react", "@testing-library/react", "../ChatBot"], factory);
  } else if (typeof exports !== "undefined") {
    factory(require("react"), require("@testing-library/react"), require("../ChatBot"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.react, global.ChatBot);
    global.undefined = mod.exports;
  }
})(this, function (_react, _react3, _ChatBot) {
  "use strict";

  var _react2 = _interopRequireDefault(_react);

  var _ChatBot2 = _interopRequireDefault(_ChatBot);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const questions = [{
    question: () => 'First Question',
    answerType: 'input',
    identifier: 'name',
    validator: {
      message: () => `Your name length should range between 5 - 26 characters`,
      validatorCallback: jest.fn()
    }
  }];
  test('chat bot render', () => {
    const onAnswer = jest.fn();

    const lastMessage = () => 'Last Message';

    const {
      getByText,
      getByTestId
    } = (0, _react3.render)( /*#__PURE__*/_react2.default.createElement(_ChatBot2.default, {
      questions: questions,
      onAnswer: onAnswer,
      lastMessage: lastMessage
    }));
    const closeBtn = getByText(/×/i);
    expect(closeBtn).toBeInTheDocument();
    closeBtn.click();
    const openButton = getByTestId('open');
    openButton.click();
  });
  test('test', () => {
    const onAnswer = jest.fn();

    const lastMessage = () => 'Last Message';

    const initialMessages = [{
      message: 'Some Random Text',
      time: '',
      sender: 'user'
    }, {
      message: 'Some File',
      fileSrc: 'https://www.youtube.com/watch?v=Dr-apUEc-7o',
      time: '',
      sender: 'user',
      file: new File(['(⌐□_□)'], 'chuck.pdf', {
        type: 'application/pdf'
      })
    }, {
      message: 'initial two',
      fileSrc: 'https://res.cloudinary.com/koech/image/upload/v1584434425/aoqxvjkdstv7fencxvok.png',
      time: '',
      sender: 'bot',
      file: new File(['(⌐□_□)'], 'chuck.png', {
        type: 'image/png'
      })
    }, {
      message: 'Some File',
      fileSrc: 'https://www.youtube.com/watch?v=Dr-apUEc-7o',
      time: '',
      sender: 'user',
      file: new File(['(⌐□_□)'], 'chuck.mp4', {
        type: 'video/mp4'
      })
    }];
    const {
      getByText
    } = (0, _react3.render)( /*#__PURE__*/_react2.default.createElement(_ChatBot2.default, {
      initialMessages: initialMessages,
      questions: [],
      onAnswer: onAnswer,
      lastMessage: lastMessage
    }));
    const message = getByText('initial two');
    expect(message).toBeInTheDocument();
  });
});
